tools/fregrid_acc/conserve_interp_acc.c

```diff
@@ -22,882 +22,606 @@
-            field_out[m].data[i] *=  (out_area[i2]/grid_out[m].cell_area[i2]);
+#pragma acc parallel loop present(out_area[:ncells_output_grid])
+        for(int i=0; i<ncells_output_grid; i++) out_area[i] = 0.0;
+
+        for(int itile=0 ; itile<ntiles_input_grid; itile++) {
+          int ncells_input_grid = input_grid[itile].nxc * input_grid[itile].nyc;
+          Interp_per_input_tile *minterp_acc = NULL ; minterp_acc = interp_acc[otile].input_tile+itile;
+          double *p_gridin_area  = NULL ; p_gridin_area  = input_grid[itile].cell_area;
+          double *p_fieldin_area = NULL ; p_fieldin_area = field_in[itile].area;
+          int ixcells = minterp_acc->nxcells;
+#pragma acc parallel loop present(minterp_acc->output_parent_cell_index[:ixcells], \
+                                  minterp_acc->input_parent_cell_index[:ixcells], \
+                                  minterp_acc->xcell_area[:ixcells],         \
+                                  out_area[:ncells_output_grid])\
+                           copyin(p_fieldin_area[:ncells_input_grid],\
+                                  p_gridin_area[:ncells_input_grid])
+          for(int ix=0; ix<ixcells; ix++) {
+            int ij2 = minterp_acc->output_parent_cell_index[ix];
+            int ij1 = minterp_acc->input_parent_cell_index[ix];
+            double area = minterp_acc->xcell_area[ix];
+            if(cell_measures ) out_area[ij2] += (area*p_fieldin_area[ij1]/p_gridin_area[ij1]);
+            else out_area[ij2] += area;
+          }
+#pragma acc parallel loop present(p_fieldout_data[:ncells_output_grid], out_area[:ncells_output_grid], \
+                                  output_grid[otile].cell_area[:ncells_output_grid])
+          for(int i=0; i<ncells_output_grid; i++) {
+            if(p_fieldout_data[i] != missing)
+              p_fieldout_data[i] *=  (out_area[i]/output_grid[otile].cell_area[i]);
           }
         }
       }
     }
 
-    free(out_area);
-    free(out_miss);
-  }
+#pragma acc exit data copyout(p_fieldout_data[:ncells_output_grid])
+#pragma acc exit data delete(out_area[:ncells_output_grid],\
+                             out_miss[:ncells_output_grid])
+
+    free(out_area); out_area = NULL;
+    free(out_miss); out_miss = NULL;
+  } // otile
 
+  return;
 
   /* conservation check if needed */
   if(opcode & CHECK_CONSERVE) {
-    double gsum_in, dd;
-    gsum_in = 0;
-    for(n=0; n<ntiles_in; n++) {
-
-      nx1  = grid_in[n].nx;
-      ny1  = grid_in[n].ny;
+    double gsum_in = 0.0;
+    for(int itile=0; itile<ntiles_input_grid; itile++) {
 
+      int nx1  = input_grid[itile].nx;
+      int ny1  = input_grid[itile].ny;
 
       if( cell_measures ) {
-        for(j=0; j<ny1; j++) for(i=0; i<nx1; i++) {
-            dd = field_in[n].data[(j+halo)*(nx1+2*halo)+i+halo];
-            if(dd != missing) gsum_in += dd*field_in[n].area[j*nx1+i];
-          }
+        for(int ij=0; ij<nx1*ny1; ij++){
+          double dd = field_in[itile].data[ij];
```

Comment:
I think it would be better to have this declared before the loop, otherwise it will need to allocate memory every time.

Reply:
Sounds good with me